### PR TITLE
Swift 4 - RxSwift 4.0 - GoogleMaps 2.5.0

### PR DESCRIPTION
**Pull request by GabrielAraujo**
This Pull request is to adjust this framework to the same layout as other RxSwiftCommunity and also to update to the latest Swift 4, RxSwift 4.0 and Google Maps 2.5.0

### CODE REVIEW

**Comment by GabrielAraujo** [comment]
I've changed the entire project structure. Due to the new way that DelegateProxy needs to be created we no longer could use some wrapper around the GMSMapView file as was with RxGMSMapView.

Other important thing to take note is that GoogleMaps still supports ios7 and due to that we can't include it into our podspec as a dependency otherwise when you try to install it will complain about static binaries.

**Comment by GabrielAraujo** [comment]
@freak4pc could you please take a look at the changes and accept the PR? We should keep the old the current project in a Version 2.3 branch.

File: Example/RxGoogleMaps/GMSCircle+Rx.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    public var position: AnyObserver<CLLocationCoordinate2D> {
+        return Binder(base) { control, position in
+            control.position = position
+            }.asObserver()
+    }
+    
+    public var radius: AnyObserver<CLLocationDistance> {
+        return Binder(base) { control, radius in
+            control.radius = radius
+            }.asObserver()
```
Comment (freak4pc) [inline Example/RxGoogleMaps/GMSCircle+Rx.swift]: Can we make sure all closing brackets are in line with the returns ?

File: Sources/GMSMapViewDelegateProxy.swift [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    open class func setCurrentDelegate(_ delegate: GMSMapViewDelegate?, to object: ParentObject) {
+        object.delegate = delegate
+    }
+    
+    public func mapView(_ mapView: GMSMapView, didTap marker: GMSMarker) -> Bool {
+        return self.didHandleTap(marker)
+    }
+    
+    func mapView(_ mapView: GMSMapView, didTap overlay: GMSOverlay) -> Void {
+        return self.didHandleTapOverlay(overlay)!
```
Comment (freak4pc) [inline Sources/GMSMapViewDelegateProxy.swift]: Can we avoid the force unwrap here?

File: Example/RxGoogleMaps/GMSMapViewDelegateProxy.swift [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    open class func setCurrentDelegate(_ delegate: GMSMapViewDelegate?, to object: ParentObject) {
+        object.delegate = delegate
+    }
+    
+    public func mapView(_ mapView: GMSMapView, didTap marker: GMSMarker) -> Bool {
+        return self.didHandleTap(marker)
+    }
+    
+    func mapView(_ mapView: GMSMapView, didTap overlay: GMSOverlay) -> Void {
+        return self.didHandleTapOverlay(overlay)!
```
Comment (freak4pc) [inline Example/RxGoogleMaps/GMSMapViewDelegateProxy.swift]: Can we please avoid the force unwrap here as well?

File: Sources/GMSMapViewDelegateProxy.swift [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        return handleMarkerInfoWindow?(marker)
+    }
+    
+    public func markerInfoContents(marker: GMSMarker) -> UIView? {
+        return handleMarkerInfoContents?(marker)
+    }
+    
+}
+
+// Referred from RxCococa.swift because it's not public
```
Comment (freak4pc) [inline Sources/GMSMapViewDelegateProxy.swift]: Can we just get rid of this comment and add a 
`// - MARK: Internal Helpers` or something of sorts

**Review by freak4pc — CHANGES_REQUESTED** [review]
Some superficial styling notes. I'll clone the project in a bit to test.

**Comment by GabrielAraujo** [comment]
@freak4pc made the corrections, I think I'm going to remove cocoapods and carthage support. We do not use them as we have to manually add the files from Sources folder into our project.
What do you think?

**Comment by freak4pc** [comment]
@GabrielAraujo Not sure I understand you? CocoaPods and Carthage support for repos is pretty required as those are the two primary methods of usage today :)

**Comment by GabrielAraujo** [comment]
@freak4pc the problem is that GoogleMaps is not a Dynamic framework and will continue like that until Google drops support for iOS7. Because of this we can't add GoogleMaps as a dependency on cocoapods. 

 This project currently use a wrapper for GoogleMaps and a bridge file to make it work but we can no longer create a wrapper around GoogleMaps as now at the DelegateProxy class you have to identify the class and the delegate. In this case `DelegateProxy<GMSMapView, GMSMapViewDelegate>`.

So because the changes I have made directly access to GoogleMaps we can't use cocoapods nor carthage as the project will always break complaining about missing GoogleMaps dependency.

**Comment by freak4pc** [comment]
That's not a good compromise to be honest. I prefer serving this as a Cocoapods/Carthage and instruct users to manually add the Google Maps dependency if there's no other choice, but completely getting rid of dependency managers is a no-go here and it's a regression from the current state of this library where you could just install and go

**Comment by freak4pc** [comment]
Also seems CocoaPods is supported? @GabrielAraujo 

030e90d7-3526-4a32-8fc7-445d05e1dd1f

**Comment by GabrielAraujo** [comment]
@freak4pc then I think the best way would keep the cocoapods installation and put a note that user must delete the files from the pod project at every pod install or update. Do you know if there is a command for automatically moving the files into the main project and deleting from RxGoogleMaps?

Regarding your second comment. You can  install directly from your pod, what you can't is to add it as a dependency in the podspec file.

You can have more details here
https://github.com/CocoaPods/CocoaPods/issues/5624

**Comment by GabrielAraujo** [comment]
@freak4pc after looking into some issues in the Cocoapods repo I discovered that they had added support for static frameworks recently. So I just had to install the latest not yet release version of cocoapods by doing `sudo gem install cocoapods --pre`
https://github.com/CocoaPods/CocoaPods/pull/6811

I think now everything is good to go!

File: RxGoogleMaps.podspec [hunk 5]
```diff
@@ -11,8 +11,10 @@ Pod::Spec.new do |s|
   s.ios.deployment_target = '8.0'
   s.source_files          = 'Sources/*.swift'
 
+  s.static_framework = true
```
Comment (freak4pc) [inline RxGoogleMaps.podspec]: Neat !

File: README.md [hunk 6]
```diff
@@ -33,13 +33,13 @@ self.view.addSubview(mapView)
 
 mapView.rx.didChangePosition.asDriver()
     .drive(onNext: { print("Did change camera position: \($0)") })
-    .addDisposableTo(disposeBag)
+    .disposed(by: disposeBag)
```
Comment (freak4pc) [inline README.md]: Except for the syntactical fixes here, are there any updates to the examples themselves that stem from your massive refactor here? Or the code should be relatively similar? Perhaps an update to the README is in order?

**Comment by freak4pc** [comment]
Hey @GabrielAraujo, There are still a few more problems that need resolution here. 

- Your Example project has a "Copy Bundle Resources" of `key.plist` but the file doesn't exist, meaning the project doesn't compile past this point. It also means I can't test the project as it requires some key. Perhaps would be good to have a demo account there with some key so it works out of the box.
515062a6-fb56-4f44-84c7-680c379c9afd

- Please update the README with some code examples to reflect the new code additions (unless you think there's nothing worth adding). Your example project seems to have MANY examples - choose the 2-3 most useful ones and put them in the README.

- Please update the README to add the following under the cocoapods Installation
```
  *Note:* Due to the fact Google Maps is delivered as a static library, you must have CocoaPods 1.4.0 installed to install this library (`gem install cocoapods --pre`)
```

Until those 3 are resolved (or the first one specifically) I have no way to functionally test this :( Let me know when you're ready. No rush!

**Comment by GabrielAraujo** [comment]
@freak4pc did the changes you asked! I just can't add the key as it would require a google account but I have added a little tutorial on how to setup the example project. =)

**Comment by freak4pc** [comment]
Hey @GabrielAraujo 
Great changes to the README. 

I went with a different approach to make the project "Compile" easily
1. I removed the plist all together,
2. I added a `YOUR-API-KEY` placeholder in AppDelegate
3. Added relevant instructions in AppDelegate and README

I also went through the code a bit deeper now and there seems to be a couple of issues: 

1. The ViewController has a bunch of commented code, If it's commented out, just delete it - make sure the file is clean with only the needed parts as part of it to reduce confusion.

2. `GMSMapViewDelegateProxy` seems to have a bunch of Closure references? This is a bit odd to me. You have a `mapView.rx.handleTapMarker = { } ` where you would expect this would be a `mapView.rx.tappedMark.bind(to: something)`. The former isn't very Rx-y, it's just a regular callback. I'm not sure why you chose this strategy, could you elaborate?

**Comment by GabrielAraujo** [comment]
@freak4pc 
1 - I removed the comments but not the code, I think that those are great examples.

2 - This was already being used previously, I tried to make Rx-y but all those functions that are using callbacks crashed at runtime. The error was something like method doesn't belong to this class. If you look at GMSMapViewDelegate you can see them there but you can't reference them as the others that are in the Rx-y way. I tried to find a solution but we can't do a deeper analysis on the GoogleMaps sdk..

**Review by freak4pc — APPROVED** [review]

**Comment by freak4pc** [comment]
Looks good to me. Let's merge this and keep improving these issues later on.

**Comment by GabrielAraujo** [comment]
Thanks @freak4pc..

We also need to update the cocoa pods page. Do you know how can we do that?

**Comment by freak4pc** [comment]
@inkyfox Can you please push a new Podspec here? If you're available to do so :)

**Pull request merged**